controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,891 @@
+		upstreamSpec.LoadBalancerSku = string(networkProfile.LoadBalancerSku)
+	}
+
+	// set linux account profile
+	linuxProfile := clusterState.LinuxProfile
+	if linuxProfile != nil {
+		upstreamSpec.LinuxAdminUsername = to.String(linuxProfile.AdminUsername)
+		sshKeys := *linuxProfile.SSH.PublicKeys
+		upstreamSpec.LinuxSSHPublicKeyContents = to.String(sshKeys[0].KeyData)
+	}
+	return upstreamSpec, err
+}
+
+func UpdateUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.SecretCache, spec aksv1.AKSClusterConfigSpec, upstreamSpec *aksv1.AKSClusterConfigSpec) (*aksv1.AKSClusterConfigSpec, error) {
+	// check kubernetes version for update
+	if spec.Tags == nil {
+		spec.Tags = map[string]string{}
+	}
+	spec.Tags["displayName"] = spec.DisplayName
+	if reflect.DeepEqual(spec.Tags, upstreamSpec.Tags) &&
+		(spec.NodePools != nil && reflect.DeepEqual(spec.NodePools, upstreamSpec.NodePools)) {
+		return nil, nil
+	}
+
+	azureAuthorizer, err := newClientAuthorizer(secretsCache, spec)
+	if err != nil {
+		return nil, err
+	}
+	resourceGroupsClient, err := newResourceGroupsClient(azureAuthorizer, spec)
+	if err != nil {
+		return nil, err
+	}
+
+	if !existsResourceGroup(ctx, resourceGroupsClient, spec) {
+		logrus.Infof("Resource group %s does not exist, creating", spec.ResourceGroup)
+		err := createResourceGroup(ctx, resourceGroupsClient, spec)
+		if err != nil {
+			return nil, fmt.Errorf("error during updating resource group %v", err)
+		}
+		logrus.Infof("Resource group [%s] updated successfully", spec)
+	}
+
+	resourceClusterClient, err := newClustersClient(azureAuthorizer, spec)
+
+	result, err := createOrUpdateCluster(ctx, secretsCache, resourceClusterClient, spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update cluster: %v", err)
+	}
+
+	clusterState := *result.ManagedClusterProperties.ProvisioningState
+	if clusterState == "Succeeded" {
+		logrus.Infof("Cluster [%s] was updated successfully", spec.DisplayName)
+	}
+
+	logrus.Infof("Cluster config [%s] was checked", spec.DisplayName)
+	return upstreamSpec, nil
+}
+
+func newClientAuthorizer(secretsCache wranglerv1.SecretCache, spec aksv1.AKSClusterConfigSpec) (autorest.Authorizer, error) {
+	authBaseURL := spec.AuthBaseURL
+	if authBaseURL == "" {
+		authBaseURL = azure.PublicCloud.ActiveDirectoryEndpoint
+	}
+	if spec.TenantID == "" {
+		return nil, fmt.Errorf("couldn't get tenantId")
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(authBaseURL, spec.TenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	clientId, clientSecret, err := getSecrets(secretsCache, spec.AzureCredentialSecret)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get secrets")
+	}
+	spToken, err := adal.NewServicePrincipalToken(*oauthConfig, clientId, clientSecret, baseURL)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't authenticate to Azure cloud")
```

Comment:
nit: consider including the error from `adal.NewServicePrincipalToken` in the message for context.